fe_utils/solvers/poisson.py

```diff
@@ -18,7 +18,78 @@ def assemble(fs, f):
     the function space in which to solve and the right hand side
     function."""
 
-    raise NotImplementedError
+
+    # Create an appropriate (complete) quadrature rule.
+    deg = fs.element.degree+2
+    cell = fs.element.cell
+    quad_rule = gauss_quadrature(cell, deg)
+
+    # Tabulate the basis functions and their gradients at the quadrature points.
+    basis_at_quad = fs.element.tabulate(quad_rule.points)
+    basis_grad_at_quad = fs.element.tabulate(quad_rule.points, grad=True)
+
+    # Create the left hand side matrix and right hand side vector.
+    # This creates a sparse matrix because creating a dense one may
+    # well run your machine out of memory!
+    A = sp.lil_matrix((fs.node_count, fs.node_count))
+    l = np.zeros(fs.node_count)
+
+    # Now loop over all the cells and assemble A and l
+
+    cell_node_map = fs.cell_nodes
+
+    num_quad_point = len(quad_rule.points)
+
+    num_of_cells = cell_node_map.shape[0]
+
+    # List of global-nodes which lie on boundary as then we know which rows of A and l to construct differently.
+    list_node_boundary = boundary_nodes(fs)
+
+    for c in range(num_of_cells):
+        # Note with our change of coordinates from cell c to reference cell, jacobian is constant in the cell.
+        # Compute some essential Jacobian related terms.
+        jacobian = fs.mesh.jacobian(c)
+        det_j = np.abs(np.linalg.det(jacobian))
+        J_inv_T = np.transpose(np.linalg.inv(jacobian))
+
+        # Number of nodes in our ref cell.
+        node_count_cell = fs.element.node_count
+
+        # vector storing the values of f at the nodes of cell c
+        f_val_in_cell_c = [f.values[cell_node_map[c, k]] for k in range(node_count_cell)]
+
+        # vector where q^{th} component stores values of f at the q^th quadrature point.
+        integral_f_cell_c = [np.dot(f_val_in_cell_c, basis_at_quad[q, :]) for q in range(num_quad_point)]
+
+        # Constructing the matrix A and l.
+        for i in range(node_count_cell):
+            # value of element in vector l.
+            integral_l = 0
+            # row_index for A and for l.
+            row_index = cell_node_map[c, i]
+            if row_index in list_node_boundary:
+                A[np.ix_(np.array([row_index]), np.array([row_index]))] = 1
```

Comment:
This approach to the boundary conditions only works because you have explicitly written out the i loop. That's really not worth the pain.